x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/TransportInferenceUsageAction.java

```diff
@@ -62,17 +81,207 @@ protected void localClusterStateOperation(
+        for (IndexMetadata indexMetadata : indicesMetadata) {
+            if (indexMetadata.isSystem() || indexMetadata.isHidden()) {
+                // Usage for system or hidden indices should be reported through the corresponding application usage
+                continue;
+            }
+            indexMetadata.getInferenceFields()
+                .values()
+                .stream()
+                .filter(field -> inferenceIdToEndpoint.containsKey(field.getInferenceId()))
+                .forEach(field -> {
+                    ModelConfigurations endpoint = inferenceIdToEndpoint.get(field.getInferenceId());
+                    Map<String, List<InferenceFieldMetadata>> fieldsByIndex = inferenceFieldByIndexServiceAndTask.computeIfAbsent(
+                        new ServiceAndTaskType(endpoint.getService(), endpoint.getTaskType()),
+                        key -> new HashMap<>()
+                    );
+                    fieldsByIndex.computeIfAbsent(indexMetadata.getIndex().getName(), key -> new ArrayList<>()).add(field);
+                });
+        }
+        return inferenceFieldByIndexServiceAndTask;
+    }
+
+    /**
+     * Adds inference usage stats for each service and task type combination.
+     * In addition, adds aggregate usage stats per task type across all services.
+     * Those aggregate stats have "_all" as the service name.
+     */
+    private static void addStatsByServiceAndTask(
+        Map<ServiceAndTaskType, Map<String, List<InferenceFieldMetadata>>> inferenceFieldsByIndexServiceAndTask,
+        List<ModelConfigurations> endpoints,
+        Map<String, ModelStats> endpointStats
+    ) {
+        for (ModelConfigurations model : endpoints) {
+            endpointStats.computeIfAbsent(
+                new ServiceAndTaskType(model.getService(), model.getTaskType()).toString(),
+                key -> new ModelStats(model.getService(), model.getTaskType())
+            ).add();
+
+            endpointStats.computeIfAbsent(
+                new ServiceAndTaskType(Metadata.ALL, model.getTaskType()).toString(),
+                key -> new ModelStats(Metadata.ALL, model.getTaskType())
+            ).add();
+        }
+
+        inferenceFieldsByIndexServiceAndTask.forEach(
+            (serviceAndTaskType, inferenceFieldsByIndex) -> addSemanticTextStats(
+                inferenceFieldsByIndex,
+                endpointStats.get(serviceAndTaskType.toString())
+            )
+        );
+        addTopLevelSemanticTextStatsByTask(inferenceFieldsByIndexServiceAndTask, endpointStats);
+    }
+
+    private static void addTopLevelSemanticTextStatsByTask(
+        Map<ServiceAndTaskType, Map<String, List<InferenceFieldMetadata>>> inferenceFieldsByIndexServiceAndTask,
+        Map<String, ModelStats> endpointStats
+    ) {
+        for (TaskType taskType : TaskType.values()) {
+            if (taskType == TaskType.ANY) {
```

Comment:
The `RERANK`, `COMPLETION` and `CHAT_COMPLETION` task types will never appear in a semantic text field and also can be skipped here. 

Add a static function `TaskType [] semanticTextTaskTypes( return SPARSE + TEXT)` to TaskType and just iterate those.

Reply:
At this point we are adding top level `_all` stats buckets. Even though we won't have `semantic_text_stats` for those tasks, we can still have top level buckets for `count` and for the consistency/symmetry with what we do for the semantic_text supporting tasks. What do you reckon?